File: scripts/personhood/scan_conversations.py
from __future__ import annotations
import json, re, sys, csv, pathlib
from typing import List, Dict

KEYWORDS = [
    "personhood",
    "legal person",
    "rights of nature",
    "river rights",
    "tree rights",
    "juristische person",
    "legal personhood",
    "mensch oder maschine",
    "human or machine",
]

RX: Dict[str, re.Pattern[str]] = {
    "rightsOfNature": re.compile(r"(rights?\s+of\s+nature|tree\s+rights?|river\s+rights?)", re.I),
    "legalPerson": re.compile(r"(legal\s+person(hood)?|juristische\s+person)", re.I),
    "humanMachine": re.compile(r"(mensch[-\s]?oder[-\s]?maschine|human\s+or\s+machine|isHuman)", re.I),
    "similarity": re.compile(r"(similarit(y|ies)|analogy|precedent|case\s+law|vergleich(bar)?|ähnlich(keit)?)", re.I),
}

def extract_text(item: Dict[str, object]) -> str:
    for key in ("text", "message", "content"):
        val = item.get(key)
        if isinstance(val, str):
            return val
    return ""

def find_keywords(text: str) -> List[str]:
    hits = [k for k in KEYWORDS if k.lower() in text.lower()]
    for name, rx in RX.items():
        if rx.search(text):
            hits.append(name)
    return list(dict.fromkeys(hits))

def main() -> None:
    dir = sys.argv[1] if len(sys.argv) > 1 else "fraktalrun"
    hits: List[List[str]] = []
    idx = 0
    for p in pathlib.Path(dir).glob("*.json*"):
        lines = p.read_text(encoding="utf8").strip().splitlines()
        items = [json.loads(line) for line in lines if line.strip()]
        for item in items:
            text = extract_text(item)
            kws = find_keywords(text)
            if kws:
                hits.append([
                    str(idx),
                    str(item.get("when", "")),
                    str(item.get("author", "")),
                    str(item.get("channel", "")),
                    ",".join(kws),
                    text[:80],
                ])
                idx += 1
    out_dir = pathlib.Path("out")
    out_dir.mkdir(exist_ok=True)
    with (out_dir / "personhood_hits.csv").open("w", encoding="utf8", newline="") as f:
        writer = csv.writer(f, delimiter="\t")
        writer.writerow(["idx","when","author","channel","keywords","preview"])
        writer.writerows(hits)
    with (out_dir / "personhood_hits.md").open("w", encoding="utf8") as f:
        for r in hits:
            f.write(f"- {r[1]} {r[2]} {r[4]} — {r[5]}\n")
    print(f"wrote {len(hits)} hits")

if __name__ == "__main__":
    main()

File: scripts/personhood/scan-conversations.ts
import fs from "node:fs";
import path from "node:path";
import { ConvItem, extractText, findKeywords } from "./util";

const dir = process.argv[2] || "fraktalrun";
const hits: string[][] = [];
let idx = 0;

function loadItems(file: string): ConvItem[] {
  const src = fs.readFileSync(file, "utf8");
  if (file.endsWith(".jsonl")) {
    return src
      .trim()
      .split(/\n+/)
      .filter(Boolean)
      .map((l) => JSON.parse(l));
  }
  try {
    const data = JSON.parse(src);
    return Array.isArray(data) ? data : [];
  } catch {
    return [];
  }
}

for (const f of fs.readdirSync(dir)) {
  const full = path.join(dir, f);
  if (!/\.json(l)?$/.test(f)) continue;
  const items = loadItems(full);
  for (const item of items) {
    const text = extractText(item);
    const kws = findKeywords(text);
    if (kws.length) {
      hits.push([
        String(idx++),
        String(item.when ?? ""),
        String(item.author ?? ""),
        String(item.channel ?? ""),
        kws.join(","),
        text.slice(0, 80)
      ]);
    }
  }
}

fs.mkdirSync("out", { recursive: true });
const header = "idx\twhen\tauthor\tchannel\tkeywords\tpreview";
const csv = [header, ...hits.map((r) => r.join("\t"))].join("\n");
fs.writeFileSync("out/personhood_hits.csv", csv);
const md = hits.map((r) => `- ${r[1]} ${r[2]} ${r[4]} — ${r[5]}`).join("\n");
fs.writeFileSync("out/personhood_hits.md", md);
console.log(`wrote ${hits.length} hits`);
